lib/puppet/provider/network/libvirt.rb

```diff
@@ -0,0 +1,177 @@
+        if ip.attribute('family') and ip.attribute('family').content == 'ipv6'
+          definition[:ipv6].push(data)
+        else
+          definition[:ip].push(data)
+        end
+      end
+    end
+    if doc.at_xpath('//mac')
+      definition[:mac] = doc.at_xpath('//mac').attribute('address').content
+    end
+    return definition
+  end
+
+  def self.instances 
+    networks = []
+    for net in $conn.list_all_networks() do
+      hash = parse_network(net)
+      hash[:autostart] = net.autostart?
+      networks << new(hash)
+    end
+    return networks
+  end
+
+  def self.prefetch(resources)
+    instances.each do |prov|
+      if resource = resources[prov.name]
+        resource.provider = prov
+      end
+    end
+  end
+
+  def create
+ #   @property_hash = @resource
+  end
+
+
+  def flush
+  debug("flushing '" + @resource[:name] + "' with: " + @property_hash.to_s )
+net_xml = <<EOF
+<network>
+  <name><%= @resource[:name] %></name>
+  <% if @property_hash[:uuid] %><uuid><%= @property_hash[:uuid] %></uuid><% end %>
+  <% if @property_hash[:mac] %>
+  <mac address='<%= @property_hash[:mac] %>'/>
+  <% end %>
+  <% if @resource[:forward_mode] %>
+  <forward<% if @resource[:forward_dev] %> dev='<%= @resource[:forward_dev] %>'<%end%> mode='<%= @resource[:forward_mode] %>'<% if !@resource[:forward_interfaces] %>/<%end%>>
+  <%  if @resource[:forward_interfaces] %>
+  <%    @resource[:forward_interfaces].each do |dev| %>
+    <interface dev='<%= dev %>'/>
+  <%    end %>
+  </forward>
+  <%  end %>
+  <% end %>
+  <% if @resource[:bridge] %>
+  <bridge name='<%= @resource[:bridge] %>'<% if @resource[:forward_mode] and @resource[:forward_mode] != 'bridge' %> stp='on' delay='0'<%end%>/>
+  <% end %>
+  <%if @resource[:ip] %>
+  <%  @resource[:ip].each do |ip| %>
+  <ip<%if ip['address']%> address='<%=ip['address']%>'<%end%><% if ip['netmask']%> netmask='<%=ip['netmask']%>'<%end%><% if ip['prefix']%> prefix='<%=ip['prefix']%>'<%end%><% unless ip['dhcp'] %>/<% end %>>
+    <% if ip['dhcp'] %>
+    <% dhcp = ip['dhcp'] %>
+    <dhcp>
+      <% if dhcp['start'] and dhcp['end']%>
+      <range start='<%=dhcp['start']%>' end='<%=dhcp['end']%>'/>
+      <%end%>
+      <% if dhcp['bootp_file']%>
+      <bootp file='<%= dhcp['bootp_file'] %>'<% if dhcp['bootp_server']%> server='<%=dhcp['bootp_server']%>'<%end%>/>
+      <%end%>
+    </dhcp>
+  </ip>
+    <% end%>
+  <%  end%>
+  <%end%>
+  <% if @resource[:ipv6] %>
+  <%  @resource[:ipv6].each do |ip| %>
+  <ip family='ipv6'<% if ip['address']%> address='<%=ip['address']%>'<%end%><% if ip['netmask']%> netmask='<%=ip['netmask']%>'<%end%><% if ip['prefix']%> prefix='<%=ip['prefix']%>'<%end%><% unless ip['dhcp'] %>/<% end %>>
+    <% if ip['dhcp'] %>
+    <% dhcp = ip['dhcp'] %>
+    <dhcp>
+      <% if dhcp['start'] and dhcp['end']%>
+      <range start='<%=dhcp['start']%>' end='<%=dhcp['end']%>'/>
+      <%end%>
+    </dhcp>
+  </ip>
+    <% end%>
+  <%  end%>
+  <%end%>
+</network>
+EOF
```

Comment:
heh. this is not pretty. actually, from what i gather, it's the same as my old template, but the whole bits around it ensure that it'll be a lot more stable.